File: planner/backend/services/atmosphere.py
"""
Atmosphere math for DCS mission planning.

Converts between ground speed (GS), calibrated airspeed (CAS), true airspeed (TAS),
and Mach number based on altitude, temperature, and wind.

DCS weather quirk: wind direction is the direction the wind BLOWS TO
(opposite of real-world meteorological convention where it's FROM).
We convert to standard (FROM) for headwind/tailwind calculations.

Standard atmosphere model (ISA):
  - Sea level temp: 15°C (288.15 K)
  - Lapse rate: -6.5°C per 1000m (up to 11km)
  - Sea level pressure: 101325 Pa
  - Speed of sound at sea level: 340.3 m/s
"""

import math
from typing import Dict, Tuple

# ISA constants
ISA_TEMP_SL = 288.15       # K (15°C)
ISA_LAPSE = 0.0065         # K/m
ISA_PRESSURE_SL = 101325   # Pa
GAMMA = 1.4                # ratio of specific heats for air
R_AIR = 287.058            # J/(kg·K) specific gas constant
G = 9.80665                # m/s²
SPEED_OF_SOUND_SL = 340.294  # m/s at sea level ISA


def isa_temperature(altitude_m: float, ground_temp_c: float = 15.0) -> float:
    """Temperature in Kelvin at altitude, using ISA lapse rate with custom ground temp."""
    temp_offset = ground_temp_c - 15.0  # deviation from ISA
    if altitude_m <= 11000:
        return (ISA_TEMP_SL + temp_offset) - ISA_LAPSE * altitude_m
    # Above tropopause — constant temp
    return (ISA_TEMP_SL + temp_offset) - ISA_LAPSE * 11000


def isa_pressure(altitude_m: float, qnh_pa: float = 101325) -> float:
    """Pressure in Pa at altitude using barometric formula."""
    if altitude_m <= 11000:
        return qnh_pa * ((ISA_TEMP_SL - ISA_LAPSE * altitude_m) / ISA_TEMP_SL) ** (G / (ISA_LAPSE * R_AIR))
    # Above tropopause
    p11 = qnh_pa * ((ISA_TEMP_SL - ISA_LAPSE * 11000) / ISA_TEMP_SL) ** (G / (ISA_LAPSE * R_AIR))
    t11 = ISA_TEMP_SL - ISA_LAPSE * 11000
    return p11 * math.exp(-G * (altitude_m - 11000) / (R_AIR * t11))


def isa_density(altitude_m: float, ground_temp_c: float = 15.0, qnh_pa: float = 101325) -> float:
    """Air density in kg/m³ at altitude."""
    p = isa_pressure(altitude_m, qnh_pa)
    t = isa_temperature(altitude_m, ground_temp_c)
    return p / (R_AIR * t)


def speed_of_sound(altitude_m: float, ground_temp_c: float = 15.0) -> float:
    """Speed of sound in m/s at altitude."""
    t = isa_temperature(altitude_m, ground_temp_c)
    return math.sqrt(GAMMA * R_AIR * t)


def tas_to_cas(tas_ms: float, altitude_m: float, ground_temp_c: float = 15.0, qnh_pa: float = 101325) -> float:
    """Convert True Airspeed to Calibrated Airspeed (both in m/s)."""
    rho = isa_density(altitude_m, ground_temp_c, qnh_pa)
    rho0 = qnh_pa / (R_AIR * ISA_TEMP_SL)  # sea level density
    if rho <= 0 or rho0 <= 0:
        return tas_ms
    return tas_ms * math.sqrt(rho / rho0)


def cas_to_tas(cas_ms: float, altitude_m: float, ground_temp_c: float = 15.0, qnh_pa: float = 101325) -> float:
    """Convert Calibrated Airspeed to True Airspeed (both in m/s)."""
    rho = isa_density(altitude_m, ground_temp_c, qnh_pa)
    rho0 = qnh_pa / (R_AIR * ISA_TEMP_SL)
    if rho <= 0 or rho0 <= 0:
        return cas_ms
    return cas_ms / math.sqrt(rho / rho0)


def tas_to_mach(tas_ms: float, altitude_m: float, ground_temp_c: float = 15.0) -> float:
    """Convert TAS to Mach number."""
    a = speed_of_sound(altitude_m, ground_temp_c)
    return tas_ms / a if a > 0 else 0


def mach_to_tas(mach: float, altitude_m: float, ground_temp_c: float = 15.0) -> float:
    """Convert Mach to TAS in m/s."""
    a = speed_of_sound(altitude_m, ground_temp_c)
    return mach * a


def gs_to_tas(gs_ms: float, heading_deg: float, altitude_m: float, wind: Dict) -> float:
    """
    Estimate TAS from ground speed, accounting for wind.

    Wind is interpolated by altitude from DCS weather layers:
      atGround, at2000, at8000

    DCS wind dir = direction wind blows TO. We convert to FROM for standard math.
    """
    wind_speed, wind_from = interpolate_wind(altitude_m, wind)

    # Headwind component (positive = headwind, slows you down)
    wind_from_rad = math.radians(wind_from)
    heading_rad = math.radians(heading_deg)
    headwind = wind_speed * math.cos(wind_from_rad - heading_rad)

    # TAS ≈ GS + headwind component
    return gs_ms + headwind


def tas_to_gs(tas_ms: float, heading_deg: float, altitude_m: float, wind: Dict) -> float:
    """Estimate ground speed from TAS, accounting for wind."""
    wind_speed, wind_from = interpolate_wind(altitude_m, wind)
    wind_from_rad = math.radians(wind_from)
    heading_rad = math.radians(heading_deg)
    headwind = wind_speed * math.cos(wind_from_rad - heading_rad)
    return tas_ms - headwind


def interpolate_wind(altitude_m: float, wind: Dict) -> Tuple[float, float]:
    """
    Interpolate wind speed (m/s) and direction (FROM, degrees) at altitude.
    DCS provides wind at ground, 2000m, 8000m with dir = TO direction.
    We convert dir to FROM (add 180°).
    """
    ground = wind.get("atGround", {})
    mid = wind.get("at2000", {})
    high = wind.get("at8000", {})

    g_spd = ground.get("speed", 0)
    g_dir = (ground.get("dir", 0) + 180) % 360  # convert TO → FROM
    m_spd = mid.get("speed", 0)
    m_dir = (mid.get("dir", 0) + 180) % 360
    h_spd = high.get("speed", 0)
    h_dir = (high.get("dir", 0) + 180) % 360

    if altitude_m <= 0:
        return g_spd, g_dir
    elif altitude_m <= 2000:
        t = altitude_m / 2000
        return _lerp(g_spd, m_spd, t), _lerp_angle(g_dir, m_dir, t)
    elif altitude_m <= 8000:
        t = (altitude_m - 2000) / 6000
        return _lerp(m_spd, h_spd, t), _lerp_angle(m_dir, h_dir, t)
    else:
        return h_spd, h_dir


def _lerp(a: float, b: float, t: float) -> float:
    return a + (b - a) * t


def _lerp_angle(a: float, b: float, t: float) -> float:
    """Interpolate between two angles (degrees), handling wraparound."""
    diff = ((b - a) + 180) % 360 - 180
    return (a + diff * t) % 360


def compute_speeds(
    gs_ms: float,
    altitude_m: float,
    heading_deg: float,
    wind: Dict,
    ground_temp_c: float = 15.0,
    qnh_pa: float = 101325,
) -> Dict:
    """
    Given ground speed, compute all speed representations.
    Returns dict with gs, tas, cas, mach (all in appropriate units).
    """
    tas = gs_to_tas(gs_ms, heading_deg, altitude_m, wind)
    cas = tas_to_cas(tas, altitude_m, ground_temp_c, qnh_pa)
    mach = tas_to_mach(tas, altitude_m, ground_temp_c)

    return {
        "gs_kts": gs_ms * 1.94384,
        "tas_kts": tas * 1.94384,
        "cas_kts": cas * 1.94384,
        "mach": round(mach, 3),
    }


def qnh_mmhg_to_pa(mmhg: float) -> float:
    """Convert QNH from mmHg to Pa."""
    return mmhg * 133.322


def qnh_mmhg_to_inhg(mmhg: float) -> float:
    """Convert QNH from mmHg to inHg."""
    return mmhg * 0.03937

File: planner/frontend/src/utils/atmosphere.ts
/**
 * Atmosphere math for speed conversions and wind calculations.
 *
 * Converts between GS (ground speed), TAS (true airspeed),
 * CAS (calibrated airspeed), and Mach based on altitude, temp, wind.
 *
 * DCS wind dir = direction wind blows TO (opposite of meteorological FROM).
 */

const ISA_TEMP_SL = 288.15;    // K (15°C)
const ISA_LAPSE = 0.0065;      // K/m
const GAMMA = 1.4;
const R_AIR = 287.058;         // J/(kg·K)
const G = 9.80665;

export interface WindLayer {
  speed: number;  // m/s
  dir: number;    // degrees (DCS convention: direction wind blows TO)
}

export interface Weather {
  wind: {
    atGround: WindLayer;
    at2000: WindLayer;
    at8000: WindLayer;
  };
  temperature_c: number;
  qnh_mmhg: number;
  qnh_inhg: number;
  qnh_hpa: number;
  clouds_base_m: number;
  clouds_preset: string;
  visibility_m: number;
}

function isaTemperature(alt_m: number, groundTempC: number = 15): number {
  const offset = groundTempC - 15;
  if (alt_m <= 11000) return (ISA_TEMP_SL + offset) - ISA_LAPSE * alt_m;
  return (ISA_TEMP_SL + offset) - ISA_LAPSE * 11000;
}

function isaPressure(alt_m: number, qnh_pa: number = 101325): number {
  if (alt_m <= 11000) {
    return qnh_pa * Math.pow((ISA_TEMP_SL - ISA_LAPSE * alt_m) / ISA_TEMP_SL, G / (ISA_LAPSE * R_AIR));
  }
  const p11 = qnh_pa * Math.pow((ISA_TEMP_SL - ISA_LAPSE * 11000) / ISA_TEMP_SL, G / (ISA_LAPSE * R_AIR));
  const t11 = ISA_TEMP_SL - ISA_LAPSE * 11000;
  return p11 * Math.exp(-G * (alt_m - 11000) / (R_AIR * t11));
}

function isaDensity(alt_m: number, groundTempC: number, qnh_pa: number): number {
  return isaPressure(alt_m, qnh_pa) / (R_AIR * isaTemperature(alt_m, groundTempC));
}

export function speedOfSound(alt_m: number, groundTempC: number = 15): number {
  return Math.sqrt(GAMMA * R_AIR * isaTemperature(alt_m, groundTempC));
}

function interpolateWind(alt_m: number, wind: Weather['wind']): { speed: number; fromDeg: number } {
  // Convert DCS "TO" direction to meteorological "FROM"
  const g = { speed: wind.atGround.speed, from: (wind.atGround.dir + 180) % 360 };
  const m = { speed: wind.at2000.speed, from: (wind.at2000.dir + 180) % 360 };
  const h = { speed: wind.at8000.speed, from: (wind.at8000.dir + 180) % 360 };

  if (alt_m <= 0) return { speed: g.speed, fromDeg: g.from };
  if (alt_m <= 2000) {
    const t = alt_m / 2000;
    return { speed: lerp(g.speed, m.speed, t), fromDeg: lerpAngle(g.from, m.from, t) };
  }
  if (alt_m <= 8000) {
    const t = (alt_m - 2000) / 6000;
    return { speed: lerp(m.speed, h.speed, t), fromDeg: lerpAngle(m.from, h.from, t) };
  }
  return { speed: h.speed, fromDeg: h.from };
}

function lerp(a: number, b: number, t: number): number { return a + (b - a) * t; }

function lerpAngle(a: number, b: number, t: number): number {
  const diff = ((b - a) + 180) % 360 - 180;
  return (a + diff * t + 360) % 360;
}

/** Estimate TAS from ground speed using wind at altitude */
export function gsToTas(gs_ms: number, heading_deg: number, alt_m: number, wind: Weather['wind']): number {
  const { speed, fromDeg } = interpolateWind(alt_m, wind);
  const headwind = speed * Math.cos((fromDeg - heading_deg) * Math.PI / 180);
  return gs_ms + headwind;
}

/** TAS to CAS */
export function tasToCas(tas_ms: number, alt_m: number, groundTempC: number, qnh_pa: number): number {
  const rho = isaDensity(alt_m, groundTempC, qnh_pa);
  const rho0 = qnh_pa / (R_AIR * ISA_TEMP_SL);
  return tas_ms * Math.sqrt(rho / rho0);
}

/** TAS to Mach */
export function tasToMach(tas_ms: number, alt_m: number, groundTempC: number): number {
  const a = speedOfSound(alt_m, groundTempC);
  return a > 0 ? tas_ms / a : 0;
}

/** Mach to TAS */
export function machToTas(mach: number, alt_m: number, groundTempC: number): number {
  return mach * speedOfSound(alt_m, groundTempC);
}

/** CAS to TAS */
export function casToTas(cas_ms: number, alt_m: number, groundTempC: number, qnh_pa: number): number {
  const rho = isaDensity(alt_m, groundTempC, qnh_pa);
  const rho0 = qnh_pa / (R_AIR * ISA_TEMP_SL);
  return cas_ms / Math.sqrt(rho / rho0);
}

/** Estimate ground speed from TAS using wind at altitude */
export function tasToGs(tas_ms: number, heading_deg: number, alt_m: number, wind: Weather['wind']): number {
  const { speed, fromDeg } = interpolateWind(alt_m, wind);
  const headwind = speed * Math.cos((fromDeg - heading_deg) * Math.PI / 180);
  return tas_ms - headwind;
}

export type SpeedMode = 'gs' | 'cas' | 'tas' | 'mach';

/**
 * Convert a pilot-entered speed in any reference to DCS ground speed (m/s).
 * This is the reverse of convertSpeed — used when the pilot edits speed.
 *
 * For GS: input is knots, convert to m/s
 * For CAS: input is knots, convert CAS→TAS→GS
 * For TAS: input is knots, convert TAS→GS
 * For Mach: input is Mach number, convert Mach→TAS→GS
 */
export function speedRefToGs(
  inputValue: number,
  mode: SpeedMode,
  alt_m: number,
  heading_deg: number,
  wx: Weather,
): number {
  const qnh_pa = wx.qnh_mmhg * 133.322;

  switch (mode) {
    case 'gs':
      return inputValue / 1.94384; // knots to m/s
    case 'cas': {
      const cas_ms = inputValue / 1.94384;
      const tas_ms = casToTas(cas_ms, alt_m, wx.temperature_c, qnh_pa);
      return tasToGs(tas_ms, heading_deg, alt_m, wx.wind);
    }
    case 'tas': {
      const tas_ms = inputValue / 1.94384;
      return tasToGs(tas_ms, heading_deg, alt_m, wx.wind);
    }
    case 'mach': {
      const tas_ms = machToTas(inputValue, alt_m, wx.temperature_c);
      return tasToGs(tas_ms, heading_deg, alt_m, wx.wind);
    }
  }
}

/**
 * Convert a stored ground speed (m/s) to the requested display mode.
 * Returns the speed in knots (or Mach number for 'mach' mode).
 */
export function convertSpeed(
  gs_ms: number,
  alt_m: number,
  heading_deg: number,
  wx: Weather,
  mode: SpeedMode,
): number {
  const qnh_pa = wx.qnh_mmhg * 133.322;
  const tas = gsToTas(gs_ms, heading_deg, alt_m, wx.wind);

  switch (mode) {
    case 'gs': return gs_ms * 1.94384;
    case 'tas': return tas * 1.94384;
    case 'cas': return tasToCas(tas, alt_m, wx.temperature_c, qnh_pa) * 1.94384;
    case 'mach': return tasToMach(tas, alt_m, wx.temperature_c);
  }
}

/**
 * Compute ETE (estimated time enroute) for a leg in seconds.
 * Uses ground speed (accounts for wind).
 */
export function computeEte(distance_m: number, gs_ms: number): number {
  return gs_ms > 0 ? distance_m / gs_ms : 0;
}

/** Format wind for display: "280°/8kts" */
export function formatWind(layer: WindLayer): string {
  const fromDir = (layer.dir + 180) % 360;
  const kts = Math.round(layer.speed * 1.94384);
  return `${Math.round(fromDir).toString().padStart(3, '0')}°/${kts}kts`;
}
